network/http_post/source/main.c

```diff
@@ -0,0 +1,183 @@
+		//ret = httpcAddPostDataAscii(&context, "data", value);
+		// If you want to add a form field containing binary data to your request, use:
+		//ret = httpcAddPostDataBinary(&context, "field name", yourBinaryData, length);
+		// If you want to add raw data to your request, use:
+		ret = httpcAddPostDataRaw(&context, (u32*)data, strlen(data));
+		printf("return from httpcAddPostDataRaw: %"PRIx32"\n",ret);
+
+		ret = httpcBeginRequest(&context);
+		if(ret!=0){
+			httpcCloseContext(&context);
+			if(newurl!=NULL) free(newurl);
+			return ret;
+		}
+
+		ret = httpcGetResponseStatusCode(&context, &statuscode);
+		if(ret!=0){
+			httpcCloseContext(&context);
+			if(newurl!=NULL) free(newurl);
+			return ret;
+		}
+
+		if ((statuscode >= 301 && statuscode <= 303) || (statuscode >= 307 && statuscode <= 308)) {
+			if(newurl==NULL) newurl = malloc(0x1000); // One 4K page for new URL
+			if (newurl==NULL){
+				httpcCloseContext(&context);
+				return -1;
+			}
+			ret = httpcGetResponseHeader(&context, "Location", newurl, 0x1000);
+			url = newurl; // Change pointer to the url that we just learned
+			printf("redirecting to url: %s\n",url);
+			httpcCloseContext(&context); // Close this context before we try the next
+		}
+	} while ((statuscode >= 301 && statuscode <= 303) || (statuscode >= 307 && statuscode <= 308));
+
+	if(statuscode!=200){
+		printf("URL returned status: %"PRIx32"\n", statuscode);
+		httpcCloseContext(&context);
+		if(newurl!=NULL) free(newurl);
+		return -2;
+	}
+
+	// This relies on an optional Content-Length header and may be 0
+	ret=httpcGetDownloadSizeState(&context, NULL, &contentsize);
+	if(ret!=0){
+		httpcCloseContext(&context);
+		if(newurl!=NULL) free(newurl);
+		return ret;
+	}
+
+	printf("reported size: %"PRIx32"\n",contentsize);
+
+	// Start with a single page buffer
+	buf = (u8*)malloc(0x1000);
+	if(buf==NULL){
+		httpcCloseContext(&context);
+		if(newurl!=NULL) free(newurl);
+		return -1;
+	}
+
+	do {
+		// This download loop resizes the buffer as data is read.
+		ret = httpcDownloadData(&context, buf+size, 0x1000, &readsize);
+		size += readsize; 
+		if (ret == (s32)HTTPC_RESULTCODE_DOWNLOADPENDING){
+				lastbuf = buf; // Save the old pointer, in case realloc() fails.
+				buf = realloc(buf, size + 0x1000);
+				if(buf==NULL){ 
+					httpcCloseContext(&context);
+					free(lastbuf);
+					if(newurl!=NULL) free(newurl);
+					return -1;
+				}
+			}
+	} while (ret == (s32)HTTPC_RESULTCODE_DOWNLOADPENDING);	
+
+	if(ret!=0){
+		httpcCloseContext(&context);
+		if(newurl!=NULL) free(newurl);
+		free(buf);
+		return -1;
+	}
+
+	// Resize the buffer back down to our actual final size
+	lastbuf = buf;
+	buf = realloc(buf, size);
+	if(buf==NULL){ // realloc() failed.
+		httpcCloseContext(&context);
+		free(lastbuf);
+		if(newurl!=NULL) free(newurl);
+		return -1;
+	}
+
+	printf("response size: %"PRIx32"\n",size);
+
+	// Print result
+	printf((char*)buf);
+
+	gfxFlushBuffers();
+	gfxSwapBuffers();
+	gspWaitForVBlank();
```

Comment:
this segment of code shoudn't be here. I think a printf(\n"); might be more appropriate just to make sure the complete contents of buf are printed.